### PR TITLE
refactor: encode metadata to be backwards compatible

### PR DESCRIPTION
**Pull request by vcastellm**
Refactoring the encoding of the values into the metadata field makes it compatible with the previously existing values, added a test for that.

Also moving tests to the types package where it belongs.

### CODE REVIEW

File: aggsender/types/types.go [hunk 1]
```diff
@@ -131,7 +131,7 @@ type CertificateMetadata struct {
 }
 
 // NewCertificateMetadataFromHash returns a new CertificateMetadata from the given hash
-func NewCertificateMetadata(fromBlock, toBlock, createdAt uint64) *CertificateMetadata {
+func NewCertificateMetadata(toBlock, fromBlock, createdAt uint64) *CertificateMetadata {
```
Comment (Stefan-Ethernal) [inline aggsender/types/types.go:134]: IMO it is more intuitive to provide the `fromBlock` as the first parameter (although we are encoding the `toBlock` first). WDYT?

```suggestion
func NewCertificateMetadata(fromBlock, toBlock, createdAt uint64) *CertificateMetadata {
```

File: aggsender/types/types.go [hunk 2]
```diff
@@ -144,24 +144,24 @@ func NewCertificateMetadataFromHash(hash common.Hash) *CertificateMetadata {
...
 }
 
 // ToHash returns the hash of the metadata
 func (c *CertificateMetadata) ToHash() common.Hash {
 	b := make([]byte, common.HashLength) // 32-byte hash
 
-	// Encode fromBlock into first 8 bytes
-	binary.BigEndian.PutUint64(b[0:8], c.FromBlock)
+	// Encode createdAt
+	binary.BigEndian.PutUint64(b[8:16], c.CreatedAt)
```
Comment (Stefan-Ethernal) [inline aggsender/types/types.go:158]: Maybe add some comment, that prior to #200, we were encoding just the `ToBlock` and that's why we are skipping the first 8 bytes...?

Although _probably_ it is worth considering versioning metadata (add the version field and act accordingly). WDYT?

Comment (vcastellm) [reply aggsender/types/types.go:158]: ok to 1.

Versioning doesn't makes much sense to me as there's only 8 bytes of space left, if we add versioning we'll use all space, no space for more changes. I think it's worth handling changes here and keeping those 8 bytes free for the future.

File: aggsender/types/types.go [hunk 3]
```diff
@@ -144,24 +144,29 @@ func NewCertificateMetadataFromHash(hash common.Hash) *CertificateMetadata {
 	b := hash.Bytes()
 
 	return NewCertificateMetadata(
-		binary.BigEndian.Uint64(b[0:8]),
-		binary.BigEndian.Uint64(b[8:16]),
 		binary.BigEndian.Uint64(b[16:24]),
+		binary.BigEndian.Uint64(b[24:32]),
+		binary.BigEndian.Uint64(b[8:16]),
 	)
```
Comment (Stefan-Ethernal) [inline aggsender/types/types.go:150]: Shouldn't this function decode the certificate metadata with the old format (namely, support the format where the `toBlock` field occupied the first 8 bytes)?

Comment (joanestebanr) [reply aggsender/types/types.go:153]: I think that requires to set the 'createAt` to `now` if it's the previous metadata version

**Review by joanestebanr — CHANGES_REQUESTED** [review]
Please, merge it at `release/v0.5.0`

File: aggsender/types/types.go (hunk 2 above)
Comment (joanestebanr) [reply aggsender/types/types.go:166]: I agree with @Stefan-Ethernal , we need to versioning metadata... We need to think in a way of doing this

**Comment by vcastellm** [comment]
> Please, merge it at `release/v0.5.0`

Instead of rebasing, let's merge it to `develop` then I'll cherry-pick to `v0.5.0`

File: aggsender/types/types.go (hunk 3 above)
Comment (vcastellm) [reply aggsender/types/types.go:153]: This is decoding @joanestebanr, so we need to take what's in `hash`.

File: aggsender/types/types.go (hunk 2 above)
Comment (vcastellm) [reply aggsender/types/types.go:166]: Do you prefer to ocuppy the remaining 8 bytes with the version?

**Comment by vcastellm** [comment]
Rebased onto `release/v0.5.0` @joanestebanr

File: aggsender/aggsender.go [hunk 4]
```diff
@@ -769,28 +774,25 @@ func extractSignatureData(signature []byte) (r, s common.Hash, isOddParity bool,
...
-
 func NewCertificateInfoFromAgglayerCertHeader(c *agglayer.CertificateHeader) *types.CertificateInfo {
 	if c == nil {
 		return nil
 	}
 	now := time.Now().UTC().UnixMilli()
+	meta := types.NewCertificateMetadataFromHash(c.Metadata)
+	toBlock := meta.FromBlock + uint64(meta.Offset)
+	if meta.Version < 1 {
+		toBlock = meta.ToBlock
```
Comment (joanestebanr) [inline aggsender/aggsender.go:787]: The `createdAt` need to be set to `now`

File: aggsender/aggsender.go [hunk 5]
```diff
@@ -769,28 +774,25 @@ func extractSignatureData(signature []byte) (r, s common.Hash, isOddParity bool,
...
 		Height:            c.Height,
 		CertificateID:     c.CertificateID,
 		NewLocalExitRoot:  c.NewLocalExitRoot,
-		FromBlock:         0,
-		ToBlock:           extractFromCertificateMetadataToBlock(c.Metadata),
+		FromBlock:         meta.FromBlock,
+		ToBlock:           toBlock,
 		Status:            c.Status,
-		CreatedAt:         now,
+		CreatedAt:         int64(meta.CreatedAt),
```
Comment (joanestebanr) [inline aggsender/aggsender.go]: maybe make sense to change the type of `CertificateInfo` for tstamp to `uint32` all of them, WDYT?

**Review by joanestebanr — APPROVED** [review]

File: aggsender/aggsender.go (hunk 5 above)
Comment (vcastellm) [reply aggsender/aggsender.go]: agree

**Review by Stefan-Ethernal — APPROVED** [review]

**Pull request merged**